docs/proposals/20220330-topology-mutation-hook.md

```diff
@@ -0,0 +1,360 @@
+
+#### Cluster topology reconciliation
+
+This section documents when the Topology Mutation Hook is going to be called during each Cluster topology reconciliation. More specifically we are going to call two different hooks for each reconciliation:
+
+* **GeneratePatches**: GeneratePatches is responsible for generating patches for the entire Cluster topology.
+* **[optional] ValidateTopology**: ValidateTopology is called after all patches have been applied and thus allows the External Patch Extension developer to validate the resulting objects.
+ 
+  **Note**: ValidateTopology is optional, i.e. it will be only called if an External Patch Extension implements it (and returns it during discovery).
+
+![Cluster topology reconciliation](./images/topology-mutation-hook/topology-reconciliation.png)
+
+#### GeneratePatches Hook
+
+A GeneratePatches call generates patches for the entire Cluster topology. Accordingly the request contains all templates, the global variables and the template-specific variables. The response contains generated patches.
+
+Example request:
+* Generating patches for a Cluster topology is done via a single call to allow External Patch Extensions a holistic view of the entire Cluster topology. Additionally this allows us to reduce the number of round-trips.
+* Each item in the request will contain the template as a raw object. Additionally information about where the template is used is provided via `holderReference`.
+```yaml
+variables:
+- name: <variable-name>
+  value: <variable-value>
+  ...
+items:
+- uid: 7091de79-e26c-4af5-8be3-071bc4b102c9
+  holderReference:
+    apiVersion: cluster.x-k8s.io/v1beta1
+    kind: MachineDeployment
+    namespace: default
+    name: cluster-md1-xyz
+    fieldPath: spec.template.spec.infrastructureRef
+  object:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+    kind: AWSMachineTemplate
+    spec:
+    ...
+  variables:
+  - name: <variable-name>
+    value: <variable-value>
+    ...
+```
+
+Example response:
+* The response contains patches instead of full objects to reduce the payload.
+* Templates in the request and patches in the response will be correlated via UIDs.
+* Like for inline patches external patches are only allowed to change `spec.template.spec`.
+```yaml
+status: Success # or Failure
+message: "error message if status == Failure"
+items:
+- uid: 7091de79-e26c-4af5-8be3-071bc4b102c9
+  patchType: JSONPatch
+  patch: <JSON-patch>
+```
+
+The full OpenAPI specification (draft) of the GeneratePatches hook can be seen via the [Swagger Editor](https://editor.swagger.io/?url=https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/docs/proposals/images/topology-mutation-hook/runtime-sdk-openapi.yaml).
```

Comment:
This link and the one below only works after merge. For review this one points to the file on the PR: https://editor.swagger.io/?url=https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/1c39879fa72a7dd77b8827f6acec2e4cca8e492a/docs/proposals/images/topology-mutation-hook/runtime-sdk-openapi.yaml